File: LightOJ/LightOJ- 1011 - Marriage Ceremonies.cpp
/*
Problrm Link: http://lightoj.com/volume_showproblem.php?problem=1011
Concept Needed: Bitwise operations, Bitmask DP.
Concept Link: Google it. There is plenty of tutorial available.
              
Learn Bitmask DP then hopefully you will be able to understand the code.
Happy coding.
*/
#include<bits/stdc++.h>
#define ll long long
#define pi 3.14159265358979323846
using namespace std;
int priority[17][17],dp[17][(1<<17)+1];
int calculate(int n, int row, int flag)
{
    if(row>=n)return 0;
    if(dp[row][flag]!=-1)return dp[row][flag];
 
    for(int i=0;i<n;i++)
        if(((1<<i)&flag)==0)
            dp[row][flag]=max(dp[row][flag],priority[row][i]+calculate(n,row+1,flag|(1<<i)));
    return dp[row][flag];
}
int main()
{
    int tc;cin>>tc;
    memset(dp,-1,sizeof(dp));
    for(int t=1;t<=tc;t++)
    {
        int n;cin>>n;
        for(int i=0;i<n;i++)
            for(int j=0;j<n;j++)
                cin>>priority[i][j];
        cout<<"Case "<<t<<": "<<calculate(n,0,1>>17)<<endl;
        memset(dp,-1,sizeof(dp));
    }
    return 0;
}

File: LightOJ/LightOJ- 1011 - Marriage Ceremonies.java
/*
Problrm Link: http://lightoj.com/volume_showproblem.php?problem=1011
Concept Needed: Bitwise operations, Bitmask DP.
Concept Link: Google it. There is plenty of tutorial available.
              
Learn Bitmask DP then hopefully you will be able to understand the code.
Happy coding.
*/
import java.util.*;
public class Main {
	
	static int point[][] = new int[16][16];
	static int dp[][] = new int[16][(1<<16)+1];
	
	static int calculate(int n, int count, int mask)
	{
		if(count>=n) return 0;
		if(dp[count][mask] != -1) return dp[count][mask];

		for(int i=0; i<n; i++)
		{
			if(((1<<i) & mask) == 0)
				dp[count][mask] = Math.max(dp[count][mask], point[count][i] + calculate(n,count+1, mask | (1<<i)) );
		}
		return  dp[count][mask];
	}
	
	public static void main(String[] args)
	{
		Scanner sc = new Scanner(System.in);
		int T;
		T = sc.nextInt();

		for(int[] row: dp)
			Arrays.fill(row, -1);

		for(int tc=1; tc<=T; tc++)
		{
			int n;
			n = sc.nextInt();
			for(int i=0; i<n; i++)
				for(int j=0;j<n;j++) point[i][j] = sc.nextInt();

			int ans = calculate(n,0,1>>17);
			System.out.println("Case "+tc+": "+ans);

			for(int[] row: dp)
				Arrays.fill(row, -1);
		}
		
	}
	
}

